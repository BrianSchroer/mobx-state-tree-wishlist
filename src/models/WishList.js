import { types, getParent, destroy } from 'mobx-state-tree';

export const WishListItem = types
  .model({
    name: types.optional(types.string, ''),
    price: types.optional(types.number, 0),
    image: '' // optional string - default to '' - shorthand for "image: types.optional(types.string, '')"
  }
  ).actions(self => ({
    changeName: newName => self.name = newName,
    changePrice: newPrice => self.price = newPrice,
    changeImage: newImage => self.image = newImage,
    remove: () => getParent(self, 2).remove(self) // 2 levels up: items, then WishList
  }));

export const WishList = types
  .model({
    items: types.optional(types.array(WishListItem), [])
  }
  ).actions(self => ({
    add: item => self.items.push(item),
    remove: item => destroy(item)
  }))
  .views(self => ({
    get totalPrice() {
      return self.items.reduce((sum, item) => sum + item.price, 0);
    }
  }));