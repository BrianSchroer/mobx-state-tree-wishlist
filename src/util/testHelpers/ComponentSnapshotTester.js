import React from 'react';
import renderer from 'react-test-renderer';

/**
 * React component snapshot helper
 */
export default class ComponentSnapshotTester {

  /**
   * Create new ComponentSnapshotTester instance
   * @param {*} component - React component
   */
  constructor(component) {
    this.component = component.type;
    this.props = component.props;
    this.adjustProps = (props, propOverrides) => Object.assign({}, props, propOverrides);
  }

  /**
   * Specify callback function to be called by the test function to adjust component
   * props with propOverrides.
   * (Defaults to adjusting top-level props.)
   */
  withPropAdjustor = (adjustor) => {
    this.adjustProps = adjustor;
    return this;
  };

  /**
   * Adjust props that were passed to the constructor with propOverrides, then create
   * snapshot and assert that it matches previously accepted screenshot.
   */
  test = (propOverrides) => {
    const Component = this.component;
    const props = this.adjustProps(this.props, propOverrides);

    const newSnapshot = renderer.create(<Component {...props} />).toJSON();

    expect(newSnapshot).toMatchSnapshot();
  }

  /**
   * Create snapshot and assert that it matches previously accepted screenshot. 
   * @param {*} component - React component1 
   */
  static test(component) {
    new ComponentSnapshotTester(component).test();
  }
}