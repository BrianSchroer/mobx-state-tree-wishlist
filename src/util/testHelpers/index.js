export { default as mobxSnapshotHelper } from './mobxSnapshotHelper/snapshotHelper';
export { default as ComponentSnapshotHelper } from './ComponentSnapshotHelper/SnapshotHelper';