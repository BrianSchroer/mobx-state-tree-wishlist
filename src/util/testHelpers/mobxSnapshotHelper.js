import { getSnapshot } from 'mobx-state-tree';

export default class {
  static createSnapshot(target) {
    return getSnapshot(target);
  }

  static assertMatch(target) {
    expect(this.createSnapshot(target)).toMatchSnapshot();
  }
}