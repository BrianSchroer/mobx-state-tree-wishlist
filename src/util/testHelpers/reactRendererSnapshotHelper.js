import renderer from 'react-test-renderer';

/**
 * react-test-renderer snapshot helper methods
 */
export default class {

  /**
   * Create component from JSX
   * @param {JSX} componentJsx (e.g. '<Button />')
   * @returns {object} component object
   */
  static createComponent(componentJsx) {
    return renderer.create(componentJsx);
  }

  /**
   * Create component from JSX and serialize as JSON "snapshot" string.
   * @param {JSX} componentJsx (e.g. '<Button />')
   * @returns {string} JSON string.
   */
  static createSnapshot(componentJsx) {
    return this.createComponent(componentJsx).toJSON();
  }

  /**
   * Create component from JSX, serialize as JSON string, and compare to accepted snapshot.
   * If a snapshot doesn't already exist, one is saved. If a previous snapshot exists
   * and the new snapshot doesn't match, you'll be asked if that's an error or if you want
   * to accept the new snapshot.
   * @param {JSX} componentJsx (e.g. '<Button />')
   */
  static assertMatch(componentJsx) {
    expect(this.createSnapshot(componentJsx)).toMatchSnapshot();
  }
}