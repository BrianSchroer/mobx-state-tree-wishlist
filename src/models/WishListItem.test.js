import { WishListItem } from './WishList';
import { mobxSnapshotHelper as snapshotHelper } from '../util/testHelpers';

const testInput = {
  name: 'Chronicles of Narnia Box Set - C.S. Lewis',
  price: 28.83,
  image: 'https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL._SX406_BO1,204,203,200_.jpg'
};

describe('WishListItem', () => {
  it('create() without input should return "empty" item', () => {
    snapshotHelper.test(WishListItem.create());
  });

  it('.create() without input should return expected values', () => {
    snapshotHelper.test(WishListItem.create(testInput));
  });

  it('.create() without image should default to empty string for image', () => {
    const input = {
      name: 'Chronicles of Narnia Box Set - C.S. Lewis',
      price: 28.83
    };

    snapshotHelper.test(WishListItem.create(input));
  });

  it('.changeName() should change name', () => {
    snapshotHelper.test(
      WishListItem.create(testInput),
      item => item.changeName('new name'));
  });

  it('.changePrice() should change price', () => {
    snapshotHelper.test(
      WishListItem.create(testInput),
      item => item.changePrice(12.34));
  });

  it('.changeImage() should change image', () => {
    snapshotHelper.test(
      WishListItem.create(testInput),
      item => item.changeImage('new image'));
  });
});