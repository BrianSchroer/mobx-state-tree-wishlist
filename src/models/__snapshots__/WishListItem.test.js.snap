// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WishListItem .changeImage() should change image 1`] = `
Array [
  Object {
    "image": "https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL._SX406_BO1,204,203,200_.jpg",
    "name": "Chronicles of Narnia Box Set - C.S. Lewis",
    "price": 28.83,
  },
  Object {
    "op": "replace",
    "path": "/image",
    "value": "new image",
  },
  Object {
    "image": "new image",
    "name": "Chronicles of Narnia Box Set - C.S. Lewis",
    "price": 28.83,
  },
]
`;

exports[`WishListItem .changeName() should change name 1`] = `
Array [
  Object {
    "image": "https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL._SX406_BO1,204,203,200_.jpg",
    "name": "Chronicles of Narnia Box Set - C.S. Lewis",
    "price": 28.83,
  },
  Object {
    "op": "replace",
    "path": "/name",
    "value": "new name",
  },
  Object {
    "image": "https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL._SX406_BO1,204,203,200_.jpg",
    "name": "new name",
    "price": 28.83,
  },
]
`;

exports[`WishListItem .changePrice() should change price 1`] = `
Array [
  Object {
    "image": "https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL._SX406_BO1,204,203,200_.jpg",
    "name": "Chronicles of Narnia Box Set - C.S. Lewis",
    "price": 28.83,
  },
  Object {
    "op": "replace",
    "path": "/price",
    "value": 12.34,
  },
  Object {
    "image": "https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL._SX406_BO1,204,203,200_.jpg",
    "name": "Chronicles of Narnia Box Set - C.S. Lewis",
    "price": 12.34,
  },
]
`;

exports[`WishListItem .create() without image should default to empty string for image 1`] = `
Object {
  "image": "",
  "name": "Chronicles of Narnia Box Set - C.S. Lewis",
  "price": 28.83,
}
`;

exports[`WishListItem .create() without input should return expected values 1`] = `
Object {
  "image": "https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL._SX406_BO1,204,203,200_.jpg",
  "name": "Chronicles of Narnia Box Set - C.S. Lewis",
  "price": 28.83,
}
`;

exports[`WishListItem create() without input should return "empty" item 1`] = `
Object {
  "image": "",
  "name": "",
  "price": 0,
}
`;
